orc8r/cloud/go/obsidian/reverse_proxy/reverse_proxy.go

```diff
@@ -50,14 +74,47 @@ func AddReverseProxyPaths(server *echo.Echo) (*echo.Echo, error) {
 				},
 			}
 			g := server.Group(prefix)
+			g.Use(r.activeBackendMiddleware)
 			g.Use(middleware.Proxy(middleware.NewRoundRobinBalancer(target)))
-			registeredPaths[prefix] = addr.String()
+			r.proxyBackendsByPathPrefix[prefix] = &reverseProxyBackend{
+				serverUrl: addr,
+				active:    true,
+			}
 		}
 	}
+	r.updateInactiveBackends(activePrefixes)
 	return server, nil
 }
 
-func getEchoServerAddressToPathPrefixes() (map[*url.URL][]string, error) {
+func (r *ReverseProxyHandler) activeBackendMiddleware(next echo.HandlerFunc) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		// Echo does not allow dynamically removing routes.
+		// If the backend is no longer active (i.e. the service no longer
+		// exists), return a 404.
+		path := c.Path()
+		path = strings.TrimSuffix(path, "/*")
+		backend, exists := r.proxyBackendsByPathPrefix[path]
+		if exists && !backend.active {
+			return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("Service not found"))
+		}
+		return next(c)
+	}
+}
+
+func (r *ReverseProxyHandler) updateInactiveBackends(activePrefixes map[string]bool) {
+	// proxyBackendsByPathPrefix contains the union of previously existing
+	// proxy configuration and all new paths that have been added. In order to
+	// figure out deactivated paths, we take the set difference of the unioned
+	// state and the current snapshot.
+	for prefix, backend := range r.proxyBackendsByPathPrefix {
+		_, exists := activePrefixes[prefix]
+		if !exists {
+			backend.active = false
+		}
+	}
+}
+
+func GetEchoServerAddressToPathPrefixes() (map[*url.URL][]string, error) {
```

Comment:
I'd like this to be an interface but since we're gonna move to a dedicated service for service discovery maybe we can table that for now